luarules/gadgets/unit_builder_priority.lua

```diff
@@ -185,101 +216,223 @@ local function UpdatePassiveBuilders(teamID, interval)
+		storage = storage * share -- consider capacity only up to the share slider
 		local reservedExpense = (resName == 'energy' and nonPassiveConsTotalExpenseEnergy or nonPassiveConsTotalExpenseMetal) -- we don't want to touch this part of expense
 		if resName == 'energy' then
-			teamStallingEnergy = cur - max(inc*stallMarginInc,stor*stallMarginSto) - 1 + (interval)*(inc-reservedExpense+rec-sent)/simSpeed --amount of res available to assign to passive builders (in next interval); leave a tiny bit left over to avoid engines own "stall mode"
+			passiveEnergyLeft = currentLevel - max(income * stallMarginInc, storage * stallMarginSto) - 1 + (income - reservedExpense + received - sent) * intervalpersimspeed --amount of res available to assign to passive builders (in next interval); leave a tiny bit left over to avoid engines own "stall mode"
 		else
-			teamStallingMetal = cur - max(inc*stallMarginInc,stor*stallMarginSto) - 1 + (interval)*(inc-reservedExpense+rec-sent)/simSpeed --amount of res available to assign to passive builders (in next interval); leave a tiny bit left over to avoid engines own "stall mode"
+			passiveMetalLeft =  currentLevel - max(income * stallMarginInc, storage * stallMarginSto) - 1 + (income - reservedExpense + received - sent) * intervalpersimspeed --amount of res available to assign to passive builders (in next interval); leave a tiny bit left over to avoid engines own "stall mode"
 		end
 	end
-
-	-- work through passive cons allocating as much expense as we have left
-	for builderID in pairs(passiveCons[teamID]) do
-		-- find out if we have used up all the expense available to passive builders yet
-		local wouldStall = false
-		if teamStallingEnergy or teamStallingMetal then
-			if passiveConsExpense[builderID] then
-				if teamStallingEnergy then
-					local passivePullEnergy = (interval*passiveConsExpense[builderID]['energy']/simSpeed)
-					if passivePullEnergy > 0 then
-						local newPullEnergy = teamStallingEnergy - passivePullEnergy
-						if newPullEnergy <= 0 then
-							wouldStall = true
-						else
-							teamStallingEnergy = newPullEnergy
+	
+	
+	-- How do we identify those units which we havent touched above yet?
+	local passiveMetalStart = passiveMetalLeft
+	local passiveEnergyStart = passiveEnergyLeft
+	
+	-- Round robining who gets a bit of the leftover resources:
+	-- This is done in two passes around the pairs(passiveConsTeam), and initially, we ignore the first roundRobinIndex units, then in second pass ignore the last roundRobinIndex units. 
+	local roundRobinIndex = roundRobinIndexTeam[teamID] or 1 -- this stores the first unit that _should_ get res
+	local havePassiveResourcesLeft = (passiveEnergyLeft > 0) and (passiveMetalLeft > 0 )
+
+	if havePassiveResourcesLeft then 
+		-- on the first pass, ignore everything < roundRobinLimit
+		-- on the second pass, ignore everything >= roundRobinLimit
+		local roundRobinLimit = roundRobinIndex
+		
+		for j=1, 2 do
+			local i = 0
```

Comment:
This is not a super helpful comment, but I find this code quite difficult to read - I don't have a good answer right now but I feel like some clearer structure or more comments might help explain what's going on. With that said, from testing I can find no problems with it.

Reply:
Yeah i can clarify